### PR TITLE
Add IMetadataImport and ISymUnmanagedSourceServerModule

### PR DESCRIPTION
**Pull request by tmat**

### CODE REVIEW

**Comment by tmat** [comment]
@gregg-miskelly Adding the one and only IMetadataImport managed COM interface def to be used in all DSR related scenarios, so that managed to managed calls work. Sounds reasonable?

**Comment by gregg-miskelly** [comment]
@tmat what scenarios are these for?

For ISymUnmanagedSourceServerModule - the only existing definition I see in VS is one for IntelliTrace. I don't think it would be hard to update, but we can double check with the IntelliTrace team.

For IMetadataImport - there may be a fair number of definitions of this. I could imagine it being somewhat tricky for people to update.

**Comment by tmat** [comment]
Re ISymUnmanagedSourceServerModule: I added a switch to our Pdb2Xml tool that displays the source server info, so I needed that interface. Last time I checked the VS debugger gets this info directly from DIA. Perhaps it would be better to use this interface to avoid direct dependency on DIA API?

Yes, there are quite a few IMetadataImport definitions -- DSR.PortablePDB also has one. Unfortunately I made it public by accident :(. The problem I run into is that when this interface is implemented in managed code and passed to managed code (DSR.PortablePDB) it blows up, since they are different managed interfaces. I though the CLR ought to unify them but that doesn't seem to be the case. Maybe I'm missing something?

**Comment by gregg-miskelly** [comment]
For IMetadataImport: If you want to get unification to work instead -- you can do one of two things --
* Stick IMetaDataImport in its own 'design time' assembly, and add the PrimaryInteropAssemblyAttribute to it. Example: https://github.com/Microsoft/MIEngine/blob/master/src/Microsoft.VisualStudio.Debugger.Interop.UnixPortSupplier/Properties/AssemblyInfo.cs
-or-
* Add the System.Runtime.InteropServices.TypeIdentifierAttribute to the interfaces and have identicial definitions in the two assemblies.

If we want this in here instead -- we should probably try and find out how much of an issue this will be for folks.

**Comment by tmat** [comment]
I would definitely not want to add another assembly.

Re the other approach with TypeIdentifierAttribute -- do the interfaces need to have exactly the same definition? If so that wouldn't help since the one already in PortablePDB doesn't have the right method signatures for some of the methods I need in the converter.

I think it would be useful to have a centralized definition that has all method definitions correctly specified. Just to avoid the pain of debugging bad interop whenever someone (including myself) makes a mistake declaring their own version.

**Comment by tmat** [comment]
BTW, I want to make the one that was mistakenly made public in PortablePDB obsolete.

File: src/Microsoft.DiaSymReader/Shared/Metadata/IMetadataImport.cs [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+namespace Microsoft.DiaSymReader
+{
+    [ComVisible(false)]
+    [ComImport]
+    [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
+    [Guid("7DAC8207-D3AE-4c75-9B67-92801A497D44")]
+    public unsafe interface IMetadataImport
+    {
+        [PreserveSig]
+        void CloseEnum(int enumHandle);
```
Comment (gregg-miskelly) [inline src/Microsoft.DiaSymReader/Shared/Metadata/IMetadataImport.cs]: Enum handles shouldn't be an int

Comment (tmat) [inline src/Microsoft.DiaSymReader/Shared/Metadata/IMetadataImport.cs]: You're right. Should be a pointer. I guess ```void*```?

Comment (gregg-miskelly) [inline src/Microsoft.DiaSymReader/Shared/Metadata/IMetadataImport.cs]: I might suggest IntPtr since it is opaque.

Comment (tmat) [inline src/Microsoft.DiaSymReader/Shared/Metadata/IMetadataImport.cs]: Yes, that's kind of same thing as void* -- it just pretends to be "safe".

**Comment by gregg-miskelly** [comment]
I agree with you that having a public definition of the IMetadata interfaces would be good. However, we do have the problem that because we are so late to define one we have lots of pre-existing definitions out there, and it is unlikely that everyone defined the interface in the same way. Also, your definition isn't a complete definition since there are several other IMetadata APIs. Lastly, the CLR team is currently considering providing their own definition.

**Comment by tmat** [comment]
>  your definition isn't a complete definition since there are several other IMetadata APIs

Yes, I know. There is IMetadataEmit, etc. But at least the Import should be complete.

> the CLR team is currently considering providing their own definition.

That would be good. I'd be happy to yield to their definition :). Who would be the right person to talk to?

**Comment by gregg-miskelly** [comment]
There are multiple 'import' interfaces also. I will start a thread.

**Comment by tmat** [comment]
Thanks!

**Comment by tmat** [comment]
Will make the interface internal.

**Pull request merged**